### PR TITLE
Fix notify_xday_threshold, implement notify_new_loans option #278

### PR DESCRIPTION
**Pull request by kryogenic**
## Description
Details are in this issue: https://github.com/Mikadily/poloniexlendingbot/issues/278

## TESTING STAGE
Working so far!

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] **I have read CONTRIBUTING.md**
- [x] **I fully understand [Github Flow.](https://guides.github.com/introduction/flow/)**
- [x] **My code adheres to the [code style of this project.](https://poloniexlendingbot.readthedocs.io/en/latest/contributing.html)**
- [x] **I have updated the documentation in /docs if I have changed the config, arguments, logic in how the bot works, or anything that understandably needs a documentation change.**
- [x] **I have updated the config file accordingly if my change requires a new configuration setting or changes an existing one.**
- [ ] **I have tested the bot with no issues for 24 continuous hours. If issues were experienced, they have been patched and tested again.**

### CODE REVIEW

File: modules/Lending.py [hunk 1]
```diff
@@ -98,6 +102,21 @@ def notify_summary(sleep_time):
     scheduler.enter(sleep_time, 1, notify_summary, (sleep_time, ))
 
 
+def notify_new_loans(sleep_time):
+    global loans_provided
+    new_provided = api.return_active_loans()['provided']
+    if loans_provided:
+        get_ids = lambda loans: map(lambda l: l['id'], loans)
```
Comment (laxdog) [inline modules/Lending.py]: I agree with Codacy here, I think this is more readable:
```
get_ids = lambda loans: [x['id'] for x in loans]
```

File: modules/Lending.py [hunk 2]
```diff
@@ -98,6 +102,21 @@ def notify_summary(sleep_time):
...
+        new_loan_ids = set(get_ids(new_provided)) - set(get_ids(loans_provided))
+        for id in new_loan_ids:
+            loan = filter(lambda l: l['id'] == id, new_provided)[0]
```
Comment (laxdog) [inline modules/Lending.py]: Up to you on this one
```
loan = [x for x in new_provided if x['id'] == id][0]
```

One thing though, what if there's more than one loan in your 60s sleep period?

Comment (kryogenic) [inline modules/Lending.py]: List comprehensions ftw! I'd take both those changes, I'm just used to using callbacks and nodejs.

> One thing though, what if there's more than one loan in your 60s sleep period?

That's what the loop is for. The filter just gets the list of all loans with a specific ID, which should always have a single element.

File: modules/Lending.py [hunk 3]
```diff
@@ -73,10 +74,13 @@ def init(cfg, api1, log1, data, maxtolend, dry_run1, analysis, notify_conf1):
 
     sleep_time = sleep_time_active  # Start with active mode
 
-    if notify_conf['enable_notifications'] and notify_conf['notify_summary_minutes']:
+    if notify_conf['enable_notifications']:
```
Comment (rnevet) [inline modules/Lending.py]: This change will make the scheduler thread run even if no notifications are configured.

Comment (kryogenic) [inline modules/Lending.py]: The thread will run but if no events are configured it will exit immediately.

Comment (kryogenic) [inline modules/Lending.py]: In retrospect we don't actually need to check `if notify_conf['enable_notifications']:` at all. To make things more clear we could check `if scheduler.empty():` before creating and starting the thread.

**Comment by kryogenic** [comment]
Ok, I've tested this merge req for 24 hours now, no bugs but I've noticed a couple oddities..

1. The notifications are far from instant, and are usually over a minute late. Strange cause the thread is supposed to run every 60 seconds. Perhaps this is due to 429 too many requests errors.

2. The scheduler thread will occasionally run into an exception and stop sending notifications. This happens more often with the patch, since we are polling `returnActiveLoans`

File: modules/Lending.py (hunk 3 above)
Comment (rnevet) [inline modules/Lending.py]: `if scheduler.empty(): `
Lets do that, keep it clear and clean.. (no body will get confused later).

**Comment by rnevet** [comment]
Regarding 2, we had the same issue in the MarketAnalysis thread - i.e. network exceptions would kill it.

That's why the main loop is surrounded by a try clause:
https://github.com/Mikadily/poloniexlendingbot/blob/master/modules/MarketAnalysis.py#L53

Could maybe use a less harsh message.

File: modules/Lending.py [hunk 4]
```diff
@@ -98,6 +102,20 @@ def notify_summary(sleep_time):
     scheduler.enter(sleep_time, 1, notify_summary, (sleep_time, ))
 
 
+def notify_new_loans(sleep_time):
+    global loans_provided
+    new_provided = api.return_active_loans()['provided']
+    if loans_provided:
+        get_id_set = lambda loans: set([x['id'] for x in loans]) # lambda to return a set of ids from the api result
+        for id in get_id_set(new_provided) - get_id_set(loans_provided):
```
Comment (rnevet) [inline modules/Lending.py]: @kryogenic what about this codacy comment?

**Comment by kryogenic** [comment]
Latest commit should cover everything.

The only problem I can see now is an increased number of `429 Too Many Requests` errors due to the excess calls to returnActiveLoans.

**Comment by rnevet** [comment]
Are you working on a shared IP? The Poloniex API implementation should prevent single instance going over the call rate limit. (You can try reducing the rate from 6 per sec.)

**Comment by rnevet** [comment]
Oh, maybe we need to make the API rate checking thread safe...

**Comment by rnevet** [comment]
I'm not sure if pass on http errors might lead to not being able to find problems....

**Comment by kryogenic** [comment]
It seems like the 429 were a result of Poloniex experiencing heavy traffic this week. I stopped getting them completely even with the notifications turned on.

For the http errors, it's up for discussion. I guess I should just follow the project conventions though and log that we're ignoring the error. Will fix.

**Comment by laxdog** [comment]
Tested this for a while now and it looks good. Nice work, this is super useful.

**Comment by rnevet** [comment]
Great, I'll merge once the exceptions are logged.

**Pull request merged**